src/tribol/mesh/MfemData.hpp

```diff
@@ -0,0 +1,1390 @@
+     * @param submesh Grid function on the (coarse) submesh
+     */
+    UpdateData(
+      const SubmeshRedecompTransfer& xfer,
+      const mfem::ParGridFunction& submesh
+    );
+    
+    /**
+     * @brief Submesh to redecomp field transfer object
+     */
+    const SubmeshRedecompTransfer& xfer_;
+
+    /**
+     * @brief Grid function values on the redecomp mesh
+     */
+    mfem::GridFunction redecomp_;
+  };
+  
+  /**
+   * @brief Get the UpdateData object
+   * 
+   * @return UpdateData& 
+   */
+  UpdateData& GetUpdateData();
+
+  /**
+   * @brief Get the UpdateData object
+   * 
+   * @return const UpdateData& 
+   */
+  const UpdateData& GetUpdateData() const;
+
+  /**
+   * @brief Grid function on the submesh
+   *
+   * @note Stored as a reference wrapper so the reference can be updated
+   */
+  std::reference_wrapper<const mfem::ParGridFunction> submesh_;
+
+  /**
+   * @brief UpdateData object created upon call to UpdateField()
+   */
+  std::unique_ptr<UpdateData> update_data_;
+};
+
+/**
+ * @brief Stores MFEM and transfer data associated with parent vector fields
+ * (displacement and velocity)
+ */
+class MfemMeshData
+{
+public:
+  /**
+   * @brief Construct a new MfemMeshData object
+   * 
+   * @param mesh_id_1 Integer identifier for first Tribol registered mesh
+   * @param mesh_id_2 Integer identifier for second Tribol registered mesh
+   * @param mesh Volume mesh of parent domain
+   * @param current_coords Grid function (on mesh) holding current coordinates
+   * @param attributes_1 Mesh boundary attributes identifying first mesh
+   * @param attributes_2 Mesh boundary attributes identifying second mesh
+   */
+  MfemMeshData(
+    integer mesh_id_1,
+    integer mesh_id_2,
+    mfem::ParMesh& mesh,
+    const mfem::ParGridFunction& current_coords,
+    const std::set<integer>& attributes_1,
+    const std::set<integer>& attributes_2
+  );
+
+  /**
+   * @brief Get coordinate grid function on the parent mesh
+   * 
+   * @return const mfem::ParGridFunction& 
+   */
+  const mfem::ParGridFunction& GetParentCoords() const
+  {
+    return coords_.GetParentField();
+  }
+
+  /**
+   * @brief Sets a new coordinate grid function on the parent mesh
+   * 
+   * @param current_coords Coordinate grid function on the parent mesh
+   */
+  void SetParentCoords(const mfem::ParGridFunction& current_coords);
+
+  /**
+   * @brief Build a new redecomp mesh and update grid functions
+   *
+   * @note This method should be called after the coordinate grid function has
+   * changed.
+   */
+  void UpdateMeshData();
+
+  /**
+   * @brief Get the integer identifier for the first Tribol registered mesh
+   *
+   * @return integer 
+   */
+  integer GetMesh1ID() const { return mesh_id_1_; }
+
+  /**
+   * @brief Get the integer identifier for the second Tribol registered mesh
+   *
+   * @return integer 
+   */
+  integer GetMesh2ID() const { return mesh_id_2_; }
+
+  /**
+   * @brief Get the number of elements on the first mesh
```

Comment:
Be clear whether this is on the first submesh, or the parent mesh, or what. What level are we at here? If the parent mesh is HO, is this the number of elements on the LOR mesh?

Reply:
Updated